### PR TITLE
Assignment5 - Alexa Darrington and Johanne McClenahan

### PR DESCRIPTION
**Pull request by JohanneJayde**
Alexa's Github - @alyxadarr
Johanne's Github - @JohanneJayde

### CODE REVIEW

**Comment by Purplepigs0129** [comment]
Reviewing

File: GenericsHomework/Node.cs [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    will be able to get back the memory and works with reference objects.
+    After some research as well, it seems the garbage collector works and handles
+    circular references and lists like our case here. 
+    The high level view of how the garbage collector works:
+    Start with locals, statics and GC pinned objects. None of these can be collected
+    Mark every object which can be reached by traversing the children of these objects Collect every object which is not marked.
+    This allows for circular references to be collected just fine. 
+    So long as none of them are reachable from an object known to be uncollectable then the circular reference is essentially irrelevant.
+    */
+    public void Clear()
```
Comment (Purplepigs0129) [inline GenericsHomework/Node.cs:60]: Great comment on the Clear method

File: GenericsTests/NodeTests.cs [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    [Fact]
+    public void ToString_NullData_ThrowsInvalidOperationException()
+    {
+        Assert.Throws<InvalidOperationException>(() => new Node<object>(null!).ToString());
+    }
+
+    [Fact]
+    public void ToString_ValidData_ReturnsStringSuccesfully()
+    {
+        Assert.Equal("Alexa", NodeList.ToString());
```
Comment (Purplepigs0129) [inline GenericsTests/NodeTests.cs:26]: I'm kind of confused about the whole generic test from what I am seeing it looks like you are but from research, it is saying to have 2 generics in the equal. With how it is written in the assignment it doesn't specify beyond having a generic in the equals. I'm going to mark it as complete for now but I could potentially be wrong but if it is then our tests our wrong too.

File: GenericsHomework/Node.cs [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+    public class Node<T>
+    {
+        public Node(T item)
+        {
+            Data = item;
+            Next = this;
+        }
+
+        public Node<T> Next { get; private set; }
```
Comment (Purplepigs0129) [inline GenericsHomework/Node.cs:12]: Definitely missed the private set on ours great job!

File: GenericsHomework/GenericsHomework.csproj [hunk 4]
```diff
@@ -0,0 +1,9 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>net8.0</TargetFramework>
+
```
Comment (Purplepigs0129) [inline GenericsHomework/GenericsHomework.csproj:5]: Same thing here could remove the whitespace

File: Directory.Build.props [hunk 5]
```diff
@@ -0,0 +1,14 @@
+<Project>
+  <!-- See https://aka.ms/dotnet/msbuild/customize for more details on customizing your build -->
+  <PropertyGroup>
+	  <Deterministic>true</Deterministic>
+	  <LangVersion>12.0</LangVersion>
+	  <ImplicitUsings>enable</ImplicitUsings>
+	  <Nullable>enable</Nullable>
+	  <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
+	  <EnableNETAnalyzers>true</EnableNETAnalyzers>
+	  <CodeAnalysisTreatWarningsAsErrors>true</CodeAnalysisTreatWarningsAsErrors>
```
Comment (Purplepigs0129) [inline Directory.Build.props:10]: Whitespace here as well unless its on purpose

File: GenericsTests/NodeTests.cs [hunk 6]
```diff
@@ -0,0 +1,87 @@
+using Xunit;
+
+namespace GenericsHomework.Tests;
+
+public class NodeTests
+{
+
+    public Node<string> NodeList { get; }
+
```
Comment (Purplepigs0129) [inline GenericsTests/NodeTests.cs:9]: Nice constructor setup for your tests

File: GenericsTests/NodeTests.cs [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+        Assert.Equal(newNode, newNode.Next);
+    }
+
+    [Fact]
+    public void Exists_NodeExistsAtEnd_ReturnsTrue()
+    {
+        Assert.True(NodeList.Exists("best!"));
+    }
+
```
Comment (Purplepigs0129) [inline GenericsTests/NodeTests.cs:66]: Good testing on seeing if all nodes still exist besides just start and end

**Review by Purplepigs0129 — COMMENTED** [review]
Create a class library project called "GenericsHomework.". ✔
Create a node class that can contain a value of any type and points to the next node and traversing the next node points back to the first item.
Define the Node class
The values should be homogeneous.
Include a constructor that takes a value. (No validation is necessary on the value). ✔
Add a ToString() override that writes out the value's ToString() result. ✔
Add a Next property that references the next node or else refers back to itself if there are no other nodes in the list. ✔
The Next property should be non-nullable (careful to follow the non-nullable property guidelines) ✔
The Next property setter should be private. ✔
Add an Append method that takes a value and appends a new Node instance after the current node (by invoking the Next property). ✔
Add a Clear method that effectively removes all items from a list except the current node. Pay attention as to whether you should be concerned with the following:
Whether it is sufficient to only set Next to itself ✔
Whether to set the removed items to circle back on themselves. In other words, whether to close the loop of the removed items. (Provide a test to show why this is required if it is required). ✔
Given there is a circular list of items, provide a comment to indicate whether you need to worry about garbage collection because all the items point to each other and therefore may never be garbage collected. ✔
Create an Exists method to test to see if a value exists in the list. ✔
Throw an error on an attempt to Append a duplicate value. (Make sure you test for this case) ✔
You should not rely on any BCL generic classes for your implementation. ✔
Extra Credit
Do one of the following two options (or both if you want extra, extra credit) :)

Implement a VennDiagram structure that contains n Circles that only contains homogenous reference types of any type. ❌
Each circle contains n items and each item can belong to one or more Circle instances.
You are not required to use a Node from earlier in the homework for your Venn diagram implementation.
You are welcome to use existing BCL generic classes for the extra credit.
Implement System.Collections.Generic.ICollection<T> on the Node class ❌
Fundamentals
Place all shared project properties into a Directory.Build.Props file.
Place all shared project items into a Directory.Build.targets file.
nullable reference types is enabled ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
and enabled .NET analyzers for both projects ✔
For this assignment, always use Assert.AreEqual<T>() (the generic version) ✔
All of the above should be unit tested ✔
Choose simplicity over complexity ✔

Didn't see any EC attempted but great job it looks like you covered everything. The only thing is if your tests are actually generic tests or some small whitespace removal. If you are going to attempt the EC best of luck! :)

File: GenericsHomework/Node.cs (hunk 1 above)
Comment (alyxadarr) [reply GenericsHomework/Node.cs:60]: thank you!

File: GenericsHomework/Node.cs (hunk 3 above)
Comment (alyxadarr) [reply GenericsHomework/Node.cs:12]: thank you! although I reviewed yours and you guys did do a private set!

**Comment by rkdeboer** [comment]
Reviewing!

File: GenericsHomework/Node.cs [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+                throw new InvalidOperationException(nameof(Data)); 
+            else
+            {
+                return Data.ToString()!;
+            }
+        }
+
+        public void Append(T item)
+        {
+            if (Exists(item))
```
Comment (rkdeboer) [inline GenericsHomework/Node.cs:27]: We also used the exists method in our append! Its a great way to reuse code for sure

File: GenericsTests/NodeTests.cs [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+        Assert.True(NodeList.Exists("Alexa"));
+    }
+
+    [Fact]
+    public void Exists_NodeDoesNotExists_ReturnsFalse()
+    {
+        Assert.False(NodeList.Exists("Rahat"));
+    }
+
+
```
Comment (rkdeboer) [inline GenericsTests/NodeTests.cs:86]: Very extensive testing nice work!

File: GenericsHomework/Node.cs [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+
+    do
+    {
+        if (currentNode.Data!.Equals(item))
+        {
+            return true;
+        }
+
+        currentNode = currentNode.Next;
+    } while (currentNode != this);
```
Comment (rkdeboer) [inline GenericsHomework/Node.cs:78]: I like the do while usage here too I did not use that when first writing this!

File: GenericsHomework/Node.cs [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+            Data = item;
+            Next = this;
+        }
+
+        public Node<T> Next { get; private set; }
+        public T Data { get; set; }
+        
+        public override string ToString()
+        {
+            if (Data==null)
```
Comment (rkdeboer) [inline GenericsHomework/Node.cs:17]: You might be able to change this to a throwifnull statement with the type of exception if you want to! This works too though!

**Review by rkdeboer — COMMENTED** [review]
Create a class library project called "GenericsHomework.". ✔
Create a node class that can contain a value of any type and points to the next node and traversing the next node points back to the first item.
Define the Node class
The values should be homogeneous.
Include a constructor that takes a value. (No validation is necessary on the value). ✔
Add a ToString() override that writes out the value's ToString() result. ✔
Add a Next property that references the next node or else refers back to itself if there are no other nodes in the list. ✔
The Next property should be non-nullable (careful to follow the non-nullable property guidelines) ✔
The Next property setter should be private. ✔
Add an Append method that takes a value and appends a new Node instance after the current node (by invoking the Next property). ✔
Add a Clear method that effectively removes all items from a list except the current node. Pay attention as to whether you should be concerned with the following:
Whether it is sufficient to only set Next to itself ✔
Whether to set the removed items to circle back on themselves. In other words, whether to close the loop of the removed items. (Provide a test to show why this is required if it is required). ✔
Given there is a circular list of items, provide a comment to indicate whether you need to worry about garbage collection because all the items point to each other and therefore may never be garbage collected. ✔
Create an Exists method to test to see if a value exists in the list. ✔
Throw an error on an attempt to Append a duplicate value. (Make sure you test for this case) ✔
You should not rely on any BCL generic classes for your implementation. ✔
Extra Credit
Do one of the following two options (or both if you want extra, extra credit) :)

Implement a VennDiagram structure that contains n Circles that only contains homogenous reference types of any type. ❌
Each circle contains n items and each item can belong to one or more Circle instances.
You are not required to use a Node from earlier in the homework for your Venn diagram implementation.
You are welcome to use existing BCL generic classes for the extra credit.
Implement System.Collections.Generic.ICollection on the Node class ❌
Fundamentals
Place all shared project properties into a Directory.Build.Props file.
Place all shared project items into a Directory.Build.targets file.
nullable reference types is enabled ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
and enabled .NET analyzers for both projects ✔
For this assignment, always use Assert.AreEqual() (the generic version) ✔
All of the above should be unit tested ✔
Choose simplicity over complexity ✔

File: GenericsTests/GenericsHomework.Tests.csproj [hunk 12]
```diff
@@ -0,0 +1,29 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>net8.0</TargetFramework>
+
```
[the remark this answers is not in the document]
Comment (alyxadarr) [reply GenericsTests/GenericsHomework.Tests.csproj:5]: will do thank you!

File: GenericsHomework/GenericsHomework.csproj (hunk 4 above)
Comment (JohanneJayde) [reply GenericsHomework/GenericsHomework.csproj:5]: Yes, we will do that :)

File: Directory.Build.props (hunk 5 above)
Comment (JohanneJayde) [reply Directory.Build.props:10]: I believe that was on purpose but would need to check again.

File: GenericsTests/NodeTests.cs (hunk 6 above)
Comment (JohanneJayde) [reply GenericsTests/NodeTests.cs:9]: Thanks! I forgot about it but then I saw I was using the same list so I did that instead

File: GenericsTests/NodeTests.cs (hunk 7 above)
Comment (JohanneJayde) [reply GenericsTests/NodeTests.cs:66]: Thanks. These tests were originally not there but then got added and helped fix bugs with checking if an item existed within the middle of node list.

File: GenericsHomework/Node.cs (hunk 8 above)
Comment (JohanneJayde) [reply GenericsHomework/Node.cs:27]: Yes!

File: GenericsTests/NodeTests.cs (hunk 9 above)
Comment (JohanneJayde) [reply GenericsTests/NodeTests.cs:86]: Thank you! I know I've had issues with doing LinkedLists in Java. Now that I know how to do tests, its very easy to check to make sure things work properly.

File: GenericsHomework/Node.cs (hunk 10 above)
Comment (JohanneJayde) [reply GenericsHomework/Node.cs:78]: I actually didn't as well until I checked another person's code and realized why it made sense to use rather than while loop

File: GenericsHomework/Node.cs (hunk 11 above)
Comment (JohanneJayde) [reply GenericsHomework/Node.cs:17]: Good catch! I need to go and fix the tests for null values and will do this :)

File: GenericsTests/NodeTests.cs (hunk 2 above)
Comment (JohanneJayde) [reply GenericsTests/NodeTests.cs:26]: Hi, Assert.Equals in the method right now is checking for strings. I can see why that is confusing since we are using generics. I will probably switch it to Assert.Equals<string> and disable warning to clarify because we are doing generics.

File: GenericsTests/VennDiagramTests.cs [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+    public void Intersection_NullFirstCircleName_ThrowsArgumentNullException()
+    {
+        Exception execption = Assert.Throws<ArgumentNullException>(() => VennDiagram.Intersection("Null Wizards", null!, "Wizards"));
+    }
+
+    [Fact]
+    public void Intersection_NullSecondCircleName_ThrowsArgumentNullException()
+    {
+        Exception execption = Assert.Throws<ArgumentNullException>(() => VennDiagram.Intersection("Null Wizards", "Wizards", null!)) ;
+    }
```
Comment (JohanneJayde) [inline GenericsTests/VennDiagramTests.cs:156]:
```suggestion
    [Fact]
    public void Intersection_NullName_ThrowsArgumentNullException()
    {

        Assert.Throws<ArgumentNullException>(() => VennDiagram.Intersection(null!, "New Yorkers", "Superheroes"));
    }

    [Fact]
    public void Intersection_NullFirstCircleName_ThrowsArgumentNullException()
    {
        Assert.Throws<ArgumentNullException>(() => VennDiagram.Intersection("Null Wizards", null!, "Wizards"));
    }

    [Fact]
    public void Intersection_NullSecondCircleName_ThrowsArgumentNullException()
    {
        Assert.Throws<ArgumentNullException>(() => VennDiagram.Intersection("Null Wizards", "Wizards", null!)) ;
    }

```
These tests don't do anything I believe. You should either Assert that Exception E is type of ArgumentNullException. This can be done using Assert.isType<ArgumentNullException>(exception). Or you just remove setting them to exception. I'd prefer dong the latter since that's how we did it in class before and on previous assignments.

File: GenericsHomework/Circle.cs [hunk 14]
```diff
@@ -0,0 +1,33 @@
+﻿
+namespace GenericsHomework;
+
+    public class Circle<T>(string Name , List<T>? items = null)
```
Comment (JohanneJayde) [inline GenericsHomework/Circle.cs:4]:
```suggestion
    public class Circle<T>(string Name)
```
By having it so you can pass in a premade list, you are creating a bug because what if you sent in a list with all duplicate values. I see that you addressed this in your VennDiagram class. It should also be addressed here.

File: GenericsHomework/Circle.cs [hunk 15]
```diff
@@ -0,0 +1,33 @@
+﻿
+namespace GenericsHomework;
+
+    public class Circle<T>(string Name , List<T>? items = null)
+    {
+        public string Name { get;} = Name ?? throw new ArgumentNullException(nameof(Name));
+        public List<T> Elements { get; private set; } = items ?? [];
```
Comment (JohanneJayde) [inline GenericsHomework/Circle.cs:7]: A way to fix the issue that I raised above would be to use a HashSet<T>. This will make so you have to only have unique values which would allow you to pass a premade set instead of only being able to set one. You also won't have to check for duplicate values within the parameter list

File: GenericsHomework/VennDiagram.cs [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+    public Circle<T> Intersection(string name, string firstCircleName, string secondCircleName)
+    {
+        ArgumentNullException.ThrowIfNull(name);
+        ArgumentNullException.ThrowIfNull(firstCircleName);
+        ArgumentNullException.ThrowIfNull(secondCircleName);
+
+        List<string> selecteCircles = [firstCircleName, secondCircleName];
+
+        List<Circle<T>> intersectCircles =
+        Circles.Where(circle => selecteCircles.Contains(circle.Name)).ToList();
```
Comment (JohanneJayde) [inline GenericsHomework/VennDiagram.cs:62]:
```suggestion
        List<Circle<T>> intersectCircles =
        Circles.Where(circle => (circle.Name.Equals( firstCircleName,StringComparison.OrdinalIgnoreCase)) || (circle.Name.Equals(secondCircleName, StringComparison.OrdinalIgnoreCase))).ToList();

```

selectedCircles could be removed and instead use two conditions with or in LINQ statement.

File: GenericsHomework/VennDiagram.cs [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+
+        //If intersection count was less than 2, that means the names are not valid
```
Comment (JohanneJayde) [inline GenericsHomework/VennDiagram.cs:64]: What if the user enters in the same name for firstCircleName and secondCircleName? It will return 1 Circle. S intersectCircles.Count will be 1. This will make it enter into the if statement where it will enter into the else statement and trigger the ArgumentException. Either handle this by making so they both can't have the same name by throwing an exception. I think it'd be best to instead just return a new Circle with the new name with elements of the one found circle.

**Review by JohanneJayde — COMMENTED** [review]
After looking back over the VennDiagram and Circle class. I found that there were issues that I did not find until after it was due.

File: GenericsTests/VennDiagramTests.cs (hunk 13 above)
Comment (BenjaminMichaelis) [reply GenericsTests/VennDiagramTests.cs:156]: I think asserting that the exception is actually the type you expect is valid. Aka, like for an argumentNullException, check that param passed into the exception is what you expect to be thrown for this

File: GenericsHomework/VennDiagram.cs [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+        foreach(Circle<T> circle in Circles[0..^1])
+        {
+            circles.Append(CultureInfo.InvariantCulture,$"{circle}, ");
+        }
+
+        circles.Append(CultureInfo.InvariantCulture, $"{Circles[^1]}");
+
+        return circles.ToString();
+    }
+    public Circle<T> Intersection(string name, string firstCircleName, string secondCircleName)
```
Comment (BenjaminMichaelis) [inline GenericsHomework/VennDiagram.cs:53]: I am not totally sure on what this `Intersection` does, it's confusing to me
